src/ladds/distributedMemParallelization/AltitudeBasedDecomposition.cpp

```diff
@@ -0,0 +1,176 @@
+  for (size_t i = 0; i < altitudeIntervals.size() - 1; i++) {
+    if (altitudeSquared >= altitudeIntervals[i] * altitudeIntervals[i] &&
+        altitudeSquared < altitudeIntervals[i + 1] * altitudeIntervals[i + 1]) {
+      return i;
+    }
+  }
+  throw std::runtime_error("Could not find rank for coordinates " + autopas::utils::ArrayUtils::to_string(coordinates));
+}
+
+std::tuple<std::array<int, 1>, std::array<int, 1>, std::array<int, 1>> LADDS::AltitudeBasedDecomposition::getGridInfo()
+    const {
+  std::array<int, 1> dims{};
+  std::array<int, 1> periods{};
+  std::array<int, 1> coords{};
+  autopas::AutoPas_MPI_Cart_get(communicator, dims.size(), dims.data(), periods.data(), coords.data());
+  return {dims, periods, coords};
+}
+
+std::vector<double> LADDS::AltitudeBasedDecomposition::logspace(const double a, const double b, const int k) {
+  double realStart = pow(10., a);
+  double realBase = pow(10., (b - a) / k);
+
+  std::vector<double> retval;
+  retval.reserve(k);
+  for (int i = 0; i < k; i++) {
+    retval.push_back(realStart);
+    realStart *= realBase;
+  }
+
+  return retval;
+}
+
+double LADDS::AltitudeBasedDecomposition::getAltitudeOfRank(const int rank) const {
+  return altitudeIntervals[rank + 1];
+}
+
+std::vector<LADDS::Particle> LADDS::AltitudeBasedDecomposition::getAndRemoveLeavingParticles(AutoPas_t &autopas) const {
+  std::vector<LADDS::Particle> particles;
+  int rank{};
+  autopas::AutoPas_MPI_Comm_rank(communicator, &rank);
+  for (auto &particle : autopas) {
+    auto logger = spdlog::get(LADDS_SPD_LOGGER_NAME);
+    SPDLOG_LOGGER_TRACE(logger.get(),
+                        "Checking particle {} at Position {} - supposed to be in rank {}",
+                        particle.getId(),
+                        autopas::utils::ArrayUtils::to_string(particle.getPosition),
+                        getRank(particle.getPosition()));
+    if (this->getRank(particle.getPosition()) != rank) {
+      particles.push_back(particle);
+      autopas.deleteParticle(particle);
+    }
+  }
+
+  return particles;
+}
+
+void LADDS::AltitudeBasedDecomposition::rebalanceDecomposition(const std::vector<LADDS::Particle> &particles,
+                                                               AutoPas_t &autopas) {
+  std::vector<double> squaredAltitudes;
+  squaredAltitudes.reserve(particles.size());
+
+  int numberOfBins{};
+  int numberOfParticles = particles.size();
+  autopas::AutoPas_MPI_Comm_size(AUTOPAS_MPI_COMM_WORLD, &numberOfBins);
+  int particlesPerRank = std::ceil(numberOfParticles / numberOfBins);
+
+  // Collect all altitudes
+  for (auto &particle : particles) {
+    squaredAltitudes.push_back(autopas::utils::ArrayMath::dot(particle.getPosition(), particle.getPosition()));
+  }
+
+  // Sort the array, then we can just get the equal partitions based on altitude from that
+  std::sort(squaredAltitudes.begin(), squaredAltitudes.end());
```

Comment:
This feels a bit like an application for [`std::partition()`](https://en.cppreference.com/w/cpp/algorithm/partition). If you have too much time on your hands feel free to test it otherwise I guess the way it is now is also fine.